src/tmt300rs/hid-tmt300rs.c

```diff
@@ -705,49 +567,44 @@ static int t300rs_update_constant(struct t300rs_device_entry *t300rs,
 	struct ff_constant_effect constant_old = old.u.constant;
 	struct __packed t300rs_packet_mod_constant {
 		struct t300rs_packet_header header;
-		uint16_t level;
+		uint16_t magnitude;
+		struct t300rs_packet_envelope envelope;
+		uint8_t effect_type;
+		uint8_t update_type;
+		uint16_t duration;
+		uint16_t offset;
 	} *packet_mod_constant = (struct t300rs_packet_mod_constant *)t300rs->send_buffer;
-	int ret;
-	int16_t level;
 
-	level = (constant.level * fixp_sin16(effect.direction * 360 / 0x10000)) / 0x7fff;
-	/* the Windows driver uses the range [-16385;16381] */
-	level = level / 2;
+	int ret = 0;
+	int16_t level, old_level;
+	uint16_t length, old_length;
 
-	if ((constant.level != constant_old.level) || (effect.direction != old.direction)) {
+	level = t300rs_calculate_constant_level(constant.level, effect.direction);
+	old_level = t300rs_calculate_constant_level(constant_old.level, old.direction);
 
-		t300rs_fill_header(&packet_mod_constant->header, effect.id, 0x0a);
-		packet_mod_constant->level = cpu_to_le16(level);
+	length = effect.replay.length - 1;
+	old_length = old.replay.length - 1;
 
-		ret = t300rs_send_int(t300rs);
-		if (ret) {
-			hid_err(t300rs->hdev, "failed modifying constant effect\n");
-			goto error;
-		}
+	if (level != old_level
+			|| t300rs_is_envelope_changed(&constant.envelope, &constant_old.envelope)
+			|| length != old_length
+			|| effect.replay.delay != old.replay.delay) {
 
-	}
+		t300rs_fill_header(&packet_mod_constant->header, effect.id, 0x6a);
+		packet_mod_constant->magnitude = cpu_to_le16(level);
 
-	ret = t300rs_update_envelope(t300rs,
-			state,
-			level,
-			effect.replay.length,
-			effect.id,
-			constant.envelope,
-			constant_old.envelope
-			);
+		t300rs_fill_envelope(&packet_mod_constant->envelope, &constant.envelope);
 
-	if (ret) {
-		hid_err(t300rs->hdev, "failed modifying constant envelope\n");
-		goto error;
-	}
+		packet_mod_constant->effect_type = 0x00;
+		packet_mod_constant->update_type = 0x45;
+		packet_mod_constant->duration = length;
+		packet_mod_constant->offset = effect.replay.delay;
```

Comment:
`->duration` and `->offset` has to take byte endianness into account, i.e. use `cpu_to_le16()`. Presumably there aren't too many people gaming on big endian systems, but still.